provider/etcd/etcd_client.go

```diff
@@ -0,0 +1,336 @@
+		return ErrorEtcdNotInitialized
+	}
+
+	fp := alert.Fingerprint()
+	alertStr, err := MarshalAlert(alert)
+	if err != nil {
+		level.Error(ec.logger).Log("msg", "Error marshaling JSON Alert", "err", err)
+		return err
+	}
+
+	// ensure the operation does not take too long
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	defer cancel()
+
+	ec.mtx.Lock()
+	_, err = ec.client.Put(ctx, ec.prefix+fp.String(), alertStr)
+	ec.mtx.Unlock()
+	if err != nil {
+		level.Error(ec.logger).Log("msg", "Error putting alert to etcd", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (ec *EtcdClient) Del(fp model.Fingerprint) error {
+	// We do a best effort.  If etcd is not initialized yet, then skip
+	if ec.client == nil {
+		level.Error(ec.logger).Log("msg", "Not deleting alert from etcd, etcd not initialized yet")
+		return ErrorEtcdNotInitialized
+	}
+
+	// ensure the operation does not take too long
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	defer cancel()
+
+	ec.mtx.Lock()
+	_, err := ec.client.Delete(ctx, ec.prefix+fp.String())
+	ec.mtx.Unlock()
+	if err != nil {
+		level.Error(ec.logger).Log("msg", "Error deleting alert from etcd", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (ec *EtcdClient) RunWatch(ctx context.Context) {
+	// watch for alert changes in etcd and writes them back to our
+	// local alert state
+	ctx = clientv3.WithRequireLeader(ctx)
+
+	go func() {
+		ec.mtx.Lock()
+		rch := ec.client.Watch(ctx, ec.prefix, clientv3.WithPrefix())
+		ec.mtx.Unlock()
+
+		for wresp := range rch {
+			for _, ev := range wresp.Events {
+				level.Debug(ec.logger).Log("msg", "watch received",
+					"type", ev.Type, "key", fmt.Sprintf("%q", ev.Kv.Key), "value", fmt.Sprintf("%q", ev.Kv.Value))
+				if ev.Type.String() == "PUT" {
+					alert, err := UnmarshalAlert(string(ev.Kv.Value))
+					if err != nil {
+						continue
+					}
+					_ = ec.alerts.Put(alert) // best effort only
+				} // else, ignore all other etcd operations, especially DELETE
+			}
+		}
+	}()
+}
+
+func (ec *EtcdClient) RunLoadAllAlerts(ctx context.Context) {
+	go func() {
+		for {
+			ec.mtx.Lock()
+			resp, err := ec.client.Get(ctx, ec.prefix, clientv3.WithPrefix())
+			ec.mtx.Unlock()
+			if err != nil {
+				level.Error(ec.logger).Log("msg", "Error fetching all alerts etcd", "err", err)
+				time.Sleep(5 * time.Second)
+				continue
+			}
+
+			for _, ev := range resp.Kvs {
+				level.Debug(ec.logger).Log("msg", "get received",
+					"key", fmt.Sprintf("%q", ev.Key), "value", fmt.Sprintf("%q", ev.Value))
+				alert, err := UnmarshalAlert(string(ev.Value))
+				if err != nil {
+					continue
+				}
+				_ = ec.alerts.Put(alert) // best effort only
+			}
+			break
+		}
+	}()
+}
+
+// Equals returns a true if the two alerts have the same values, false otherwise
+func AlertsEqualExceptForUpdatedAt(a *types.Alert, o *types.Alert) bool {
```

Comment:
MAY: a single `return` with a long logical expression is clearer to me.

Reply:
Done, in next commit